Document/0x06c-Reverse-Engineering-and-Tampering.md

```diff
@@ -609,6 +609,379 @@ Start Safari on the iOS device. Run the above Python script on your connected ho
 
+##### Process Exploration
+
+When testing an app, process exploration can provide the tester with deep insights into the app process memory. It can be achieved via runtime instrumentation and allows to perform tasks such as:
+
+- Retrieving the memory map and loaded libraries.
+- Searching for occurrences of certain data.
+- After doing a search, obtaining the location of a certain offset in the memory map.
+- Performing a memory dump and inspect or reverse engineer the binary data _offline_.
+- Reverse engineering a binary or Framework while it's running.
+
+As you can see, these tasks are rather supportive and/or passive, they'll help us collect data and information that will support other techniques. Therefore, they're normally used in combination with other techniques such as method hooking.
+
+In the following sections you will be using r2frida to retrieve information straight from the app runtime. First start by opening an r2frida session to the target app (e.g. iGoat-Swift) that should be running on your iPhone (connected per USB). Use the following command:
+
+```bash
+$ r2 frida://usb//iGoat-Swift
+```
+
+> See all options with `r2 frida://?`.
+
+Once in the r2frida session, all commands start with `\`. For example, in radare2 you'd run `i` to display the binary information, but in r2frida you'd use `\i`.
+
+###### Memory Maps and Inspection
+
+You can retrieve the app's memory maps by running `\dm`:
+
+```bash
+[0x00000000]> \dm
+0x0000000100b7c000 - 0x0000000100de0000 r-x /private/var/containers/Bundle/Application/3ADAF47D-A734-49FA-B274-FBCA66589E67/iGoat-Swift.app/iGoat-Swift
+0x0000000100de0000 - 0x0000000100e68000 rw- /private/var/containers/Bundle/Application/3ADAF47D-A734-49FA-B274-FBCA66589E67/iGoat-Swift.app/iGoat-Swift
+0x0000000100e68000 - 0x0000000100e97000 r-- /private/var/containers/Bundle/Application/3ADAF47D-A734-49FA-B274-FBCA66589E67/iGoat-Swift.app/iGoat-Swift
+...
+0x0000000100ea8000 - 0x0000000100eb0000 rw-
+0x0000000100eb0000 - 0x0000000100eb4000 r--
+0x0000000100eb4000 - 0x0000000100eb8000 r-x /usr/lib/TweakInject.dylib
+0x0000000100eb8000 - 0x0000000100ebc000 rw- /usr/lib/TweakInject.dylib
+0x0000000100ebc000 - 0x0000000100ec0000 r-- /usr/lib/TweakInject.dylib
+0x0000000100f60000 - 0x00000001012dc000 r-x /private/var/containers/Bundle/Application/3ADAF47D-A734-49FA-B274-FBCA66589E67/iGoat-Swift.app/Frameworks/Realm.framework/Realm
+```
+
+While you're searching or exploring the app memory, you can always verify where you're located in each moment (where your current offset is located) in the memory map. Instead of noting and searching for the memory address in this list you can simply run `\dm.`. You'll find an example in the following section "In-Memory Search".
```

Comment:
```suggestion
While you're searching or exploring the app memory, you can always verify where your current offset is located in the memory map. Instead of noting and searching for the memory address in this list you can simply run `\dm.`. You'll find an example in the following section "In-Memory Search".
```